src/transactions.rs

```diff
@@ -1553,14 +1553,20 @@ impl WriteTransaction {
         user_root: Option<BtreeHeader>,
         eventual: bool,
     ) -> Result {
+        // If there are no active read transactions, we can free pages from the current transaction
         let free_until_transaction = self
             .transaction_tracker
             .oldest_live_read_transaction()
-            .map_or(self.transaction_id, |x| x.next());
+            .map_or(self.transaction_id.next(), |x| x.next());
+
+        // Free pages from previous transactions before we save the allocator state
+        // This ensures the freed pages are available for reuse in the next transaction
         self.process_freed_pages(free_until_transaction)?;
 
+        self.process_freed_pages(self.transaction_id)?;
```

Comment:
I think you need to remove this. It won't be safe to do frees from the current transaction because we might crash before the fsync goes through, in which case the database will roll back